src/webhook/regular_user_validation.py

```diff
@@ -25,22 +31,30 @@ def handle_request():
     return responses.response_invalid()
   
   try:
+    body_dict = request.json['request']
+    username = body_dict['userInfo']['username']
+
+    if body_dict['object'] is None:
+      group_name = body_dict['oldObject']['metadata']['name']
+    else:
+      group_name = body_dict['object']['metadata']['name']
+
     # get the username and decide if it's a special user (SA, kube:admin, etc) or not.. 
     # deny any request if it's not a special user:  kube:*, system:* except system:unauthenticated
+    # ALLOW any request if the group_name is in admin_groups
     # reference: https://kubernetes.io/docs/reference/access-authn-authz/extensible-admission-controllers/#request
-    body_dict = request.json['request']
-    username = body_dict['userInfo']['username']
-    if username == "system:unauthenticated" or (not username.startswith("kube:") and not username.startswith("system:")):
+    if is_request_allowed(username, group_name, admin_groups):
       kind = body_dict['object']['kind']
       operation = body_dict['operation']
-      namespace = None
-      if 'namespace' in body_dict:
-        namespace = body_dict['namespace']
       return responses.response_deny(req=body_dict, msg="Regular user '{}' cannot {} kind '{}'.".format(username, operation, kind))
     else:
       return responses.response_allow(req=body_dict)
   except Exception:
     print("Exception:")
     print("-"*60)
     traceback.print_exc(file=sys.stdout)
-    return responses.response_invalid()
\ No newline at end of file
+    return responses.response_invalid()
+
+
+def is_request_allowed(username, groupname, admin_groupnames=[]):
+  return not(groupname not in admin_groupnames and (username == "system:unauthenticated" or (not username.startswith("kube:") and not username.startswith("system:"))))
```

Comment:
Yeah, this is way hard to understand IMO, and very not-pythonic.
Suggestion part 3:
```suggestion
  # Explicitly unauthenticated
  if username == "system:unauthenticated":
      return False
  if groupname in admin_groupnames:
      return True
  if username.startswith("kube:") or username.startswith("system:"):
      return True
  # Deny anyone else
  return False
```
If you do this, you don't need the comments removed from above at all:
```
    # deny any request if it's not a special user:  kube:*, system:* except system:unauthenticated
    # ALLOW any request if the group_name is in admin_groups
```
...because that's obvious from the code.